game-core/src/main/java/games/strategy/engine/framework/ui/SaveGameFileChooser.java

```diff
@@ -1,48 +1,56 @@
 
 /** A file chooser for save games. Defaults to the user's configured save game folder. */
 public final class SaveGameFileChooser extends JFileChooser {
   private static final long serialVersionUID = 1548668790891292106L;
 
-  private static SaveGameFileChooser instance;
-
-  private SaveGameFileChooser() {
-    setFileFilter(newGameDataFileFilter());
-    final File saveGamesFolder = ClientSetting.saveGamesFolderPath.getValueOrThrow().toFile();
-    ensureDirectoryExists(saveGamesFolder);
-    setCurrentDirectory(saveGamesFolder);
-  }
-
-  public static SaveGameFileChooser getInstance() {
-    if (instance == null) {
-      instance = new SaveGameFileChooser();
+  /**
+   * Displays a file chooser dialog for the user to select the file to which the current game should
+   * be saved.
+   *
+   * @param frame The owner of the file chooser dialog; may be {@code null}.
+   * @return The file to which the current game should be saved or {@code null} if the user
+   *     cancelled the operation.
+   */
+  public static File getSaveGameLocation(final Frame frame, final GameData gameData) {
+    final FileDialog fileDialog = new FileDialog(frame);
+    fileDialog.setMode(FileDialog.SAVE);
+    fileDialog.setDirectory(ClientSetting.saveGamesFolderPath.getValueOrThrow().toString());
+    fileDialog.setFilenameFilter((dir, name) -> GameDataFileUtils.isCandidateFileName(name));
+    fileDialog.setFile(getSaveGameName(gameData));
+
+    fileDialog.setVisible(true);
+    final String fileName = fileDialog.getFile();
+    if (fileName == null) {
+      return null;
     }
-    return instance;
-  }
 
-  private static void ensureDirectoryExists(final File f) {
-    if (!f.mkdirs() && !f.exists()) {
-      throw new IllegalStateException("Unable to create save game folder: " + f.getAbsolutePath());
-    }
+    // If the user selects a filename that already exists,
+    // the AWT Dialog will ask the user for confirmation
+    return new File(fileDialog.getDirectory(), GameDataFileUtils.addExtensionIfAbsent(fileName));
   }
 
-  private static FileFilter newGameDataFileFilter() {
-    return new FileFilter() {
-      @Override
-      public boolean accept(final File file) {
-        return file.isDirectory() || GameDataFileUtils.isCandidateFileName(file.getName());
-      }
+  private static String getSaveGameName(final GameData gameData) {
+    return gameData.getSaveGameFileName().orElse(formatGameName(gameData.getGameName()));
+  }
 
-      @Override
-      public String getDescription() {
-        return "Saved Games, *" + GameDataFileUtils.getExtension();
-      }
-    };
+  private static String formatGameName(final String gameName) {
+    final ZonedDateTime now = ZonedDateTime.now();
+    return new StringJoiner("-")
+        .add(String.valueOf(now.getYear()))
+        .add(String.valueOf(now.getMonthValue()))
+        .add(String.valueOf(now.getDayOfMonth()))
+        .add(gameName.replaceAll(" ", "-") + GameDataFileUtils.getExtension())
```

Comment:
Thoughts about using https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/format/DateTimeFormatter.html#ISO_LOCAL_DATE for date formatting?

Reply:
It's better to use built-in items in general, but primitive and simple has it's virtue as well. IMO it's a wash here, there is nothing difficult to understand about this code. Meanwhile there is a format code that ties the year to the calendar week, which causes Dec-31 to flip to Jan-1, so there is some subtlety/caution to be used there.